### PR TITLE
Several ffi_support improvements

### PR DESCRIPTION
**Pull request by thomcc**
- The main one here is the addition of the `FfiStr` type. This allows most of our ffi code to be written without any `unsafe`, unless it's doing something complex (in which case indicating this with `unsafe` is appropriate). The old functions are still present, but deprecated, mostly for documentation purposes.
- `ffi_support` now properly handles unwind-safety. In practice we already handle this properly, as almost everything uses HandleMaps now-a-days, so there's really no reason to default to the panic-unsafe thing. This also lets us remove a big block comment defending our current choice to do the dubious thing.
- It removes explicit dependencies on `serde`/`serde_json`/`prost` from `ffi_support` when they're only used in macros and the caller already needs dependencies on them for them for them to work. This removes the `prost_support` feature, since now those macros are just always available.
- And while I was at it: I've also removed the ability to compile the components without their `ffi.rs` module which generally jsut implements IntoFfi. Making this a feature causes unnecessary recompilation, and ffi_support is quite lightweight, so having the dependency doesn't hurt us at all.

In general, these changes should make things safer and build faster. The first two have been somethign I've wanted to do for a while, but submitting to rustconf made me remember that I should clean this up.

### CODE REVIEW

**Comment by vladikoff** [comment]
> The main one here is the addition of the FfiStr type. This allows most of our ffi code to be written without any unsafe, unless it's doing something complex (in which case indicating this with unsafe is appropriate). The old functions are still present, but deprecated, mostly for documentation purposes.

That's awesome! 🔥 🔥 🔥 🔥

File: components/support/ffi/src/lib.rs [hunk 1]
```diff
@@ -87,20 +88,22 @@
...
+//! which contains several helpful inherent methods for extracting their data.
 //!
 
 use std::{panic, thread};
 
 mod error;
+mod ffistr;
 pub mod handle_map;
 mod into_ffi;
+#[macro_use]
```
Comment (eoger) [inline components/support/ffi/src/lib.rs]: I assume this is still needed despite the `crate::macros::*;` import bellow

File: components/fxa-client/ffi/src/lib.rs [hunk 2]
```diff
@@ -34,17 +34,17 @@ lazy_static::lazy_static! {
 /// A destructor [fxa_free] is provided for releasing the memory for this
 /// pointer type.
 #[no_mangle]
-pub unsafe extern "C" fn fxa_new(
-    content_url: *const c_char,
-    client_id: *const c_char,
-    redirect_uri: *const c_char,
+pub extern "C" fn fxa_new(
+    content_url: FfiStr<'_>,
```
Comment (eoger) [inline components/fxa-client/ffi/src/lib.rs]: Looks awesome

File: components/support/ffi/src/lib.rs (hunk 1 above)
Comment (thomcc) [reply components/support/ffi/src/lib.rs]: Yeah, seems to be. Not sure what the deal is, tbh

**Pull request merged**